README.md
# progress-scroll 滚动进度可视化插件

> 🤖🎉🎉 进度监控可视化插件是一个基于 原生 `Css + Vue3 钩子` 封装的实用 hooks 工具，用于在 Vue.js 应用程序中展示进度。它提供了一种简单、高效的操作体验、深深感受开发的乐趣 ~

## 📦 体验

[一键速看](https://huozaifenlangli.github.io/Vue3-template/#/scroll)

## 🌍 安装

你可以使用 npm 或 pnpm 安装插件：

```javascript
npm i vue3-progress-scroll
```

## 🛹 注入

在你的主应用程序入口文件（例如 main.js）中，导入并使用 ：
**_`main.js`_**

```javascript
/* Step 1
------------------------------------------------------------------ */
import { useScroll } from "vue3-progress-scroll";
const app = createApp(App);
app.use(useScroll);
```

### 🎉 配置

```javascript
/* Step 2
------------------------------------------------------------------ */
// 注册插件并提供自定义的进度条属性（可选）例如：
app.use(useScroll, {
	progress: "green", // 进度条颜色
	progressRollback: "#fff", // 进度条回滚颜色
	progressTop: "3px", // 进度条距离顶部的距离
	progressLeft: "0px", // 进度条距离左边的距离
	UIViewBackground: "#fff", // 页面背景色
});
```

## 🤖 使用方法

一旦你设置了插件，你就可以在组件中使用 $openScroll 和 $closeScroll() 方法：

> 第一种 inject

```javascript
/* Step 3
------------------------------------------------------------------ */
// 💡 在父容器绑定类名
<div class="ProgressTopBar">
    <p v-for="(item, index) in 1000" :key="index">{{ index + 1 }}</p>
</div>

import { ref, inject } from "vue";
import { scrollKey } from "vue3-progress-scroll";
inject<typeof scrollKey>(scrollKey).$openScroll();
inject<typeof scrollKey>(scrollKey).$closeScroll();
```

> 第二种 getCurrentInstance()

```javascript
import { getCurrentInstance } from "vue";
const { proxy }: any = getCurrentInstance();
proxy.$openScroll();
proxy.$closeScroll();
```

### 📝 使用示例 Demo.vue

```javascript

<template>
	<div class="ProgressTopBar">
		<p v-for="(item, index) in items" :key="index">{{ item }}</p>
	</div>
</template>

<script setup lang="ts">
	// vue3.2.0+ 之后支持的新特性
	defineOptions({
		name: "scrollBar", // 组件名
		inheritAttrs: false, // 是否继承父组件的属性
	});
	import { ref, inject } from "vue";
	const items = ref<string[]>([]);
	for (let i = 0; i < 200; i++) items.value.push(`Item ${i + 1}`);

   import { scrollKey } from "vue3-progress-scroll";
   inject<typeof scrollKey>(scrollKey).$openScroll();
   inject<typeof scrollKey>(scrollKey).$closeScroll();
</script>

```

## 💌 原理

[一键快阅源码（约 20 行）9-28](https://zhang-kun8888.gitee.io/vue3-tools-docs/vue3-progress-scroll.html#%F0%9F%92%8C-%E5%8E%9F%E7%90%86)


CHANGELOG.md
# <a href="https://github.com/huozaifenlangli/vue3-progress-scroll--NPM" target="_blank">vue3-progress-scroll 更新日志</a>

🎉🎉🤖 `vue3-progress-scroll` 轻量版 滚动进度可可视化插件, 融入 Vue3 + Ts Hooks, 更方便使用

## 1.2.X

`2023.10.26`

- 🌟 更新 上传 1.2.6 版本
- 🎉 更新编译方式

---

## 1.2.X

`2023.10.09`

- 🌟 更新 上传 1.2.5 版本
- 🎉 修复 `lib/index` 无参数处理

---

## 1.1.X

`2023.10.09`

- 🌟 更新 上传 1.1.6 版本
- 🎉 更新 `lib/index` 滚动天距离距离顶部位置

---

- 🌟 更新 上传 1.1.4 版本
- 🎉 更新 `package.json URL` 文档

---

- 🌟 更新 上传 1.1.3 版本
- 🌟 更新 `README.md` 文档

## 1.0.X

`2023.08.07`

- 🌟 更新 上传 1.0.4 版本
- 🐞 修复 package.json 文件

---

- 🌟 更新 上传 1.0.3 版本
- 🎉 新增 `/tsconfig.json` `/dist` 编译 Ts 并使用 js 上传 npm，更好的兼容没使用 ts 的 Vue 项目

---

- 🌟 更新 上传 1.0.2 版本
- 🐞 修复 `/utils/tools` 判断加固
- 🐞 修复 -

---

- 🌟 更新 上传 1.0.1 版本
- 🐞 修复 `/directive/index` 未导出错误
- 🐞 修复 -

---

- 🌟 更新 上传 1.0.0 版本
- 🎉 新增 `/utils/tools` 判断奇数偶数功能
- 🐞 修复 -
- 🐞 修复 -


lib/index.ts
import { App, onActivated, onDeactivated, onMounted, onUnmounted } from "vue";
import type { InjectionKey } from "vue";

interface ScrollBarType {
	progress: string;
	progressRollback: string;
	progressTop: string;
	progressLeft: string;
	UIViewBackground: string;
}

let timer: number | null | undefined | NodeJS.Timeout = null;
// 声明provide的key用于inject接受类型注解
const scrollKey: InjectionKey<{
	// 参数可选、因为只需要在main.ts中传入一次参数即可，其余地方调用不需要传参
	$openScroll: (option?: ScrollBarType) => void;
	$closeScroll: () => void;
}> = Symbol();

function onUnSetProgressProp(): void {
	const el = document.querySelector<HTMLElement>(".ProgressTopBar");

	el?.style?.removeProperty("background-image");
	el?.style?.removeProperty("background-size");
	el?.style?.removeProperty("background-repeat");
}

function onSetProgressProp(
	option: ScrollBarType = {
		progress: "red",
		progressRollback: "#fff",
		progressTop: "3px",
		progressLeft: "0px",
		UIViewBackground: "#fff",
	}
): void {
	timer = setTimeout(() => {
		const el = document.querySelector(".ProgressTopBar");
		if (el && el instanceof HTMLElement) {
			el.style.setProperty("position", "relative");
			el.style.setProperty("z-index", "1");
			el.style.setProperty(
				"background-image",
				`linear-gradient(to right top,  ${option.progress} 50%, ${option.progressRollback} 50%)`
			);
			el.style.setProperty(
				"background-size",
				`100% calc(100% - 100vh + ${option.progressTop} )`
			);
			el.style.setProperty("background-repeat", "no-repeat");
			//  无法设置CSS StyleDeclaration 的背景颜色属性：这些样式是计算出来的，因此背景颜色属性是只读的
			// console.log((getComputedStyle(el, '::after').backgroundColor = 'yellow'));
			const styleEl = document.createElement("style");
			styleEl.innerHTML = `
				.ProgressTopBar::after {
					content: '';
					position: fixed;
					top: ${option.progressTop};
					left: ${option.progressLeft};
					bottom: 0;
					right: 0;
					background: ${option.UIViewBackground};
					z-index: -1;
				}
			`;
			document.head.appendChild(styleEl);
		}
	}, 500);
}

// 命名函数openScroll  参数可选、因为只需要在main.ts中传入一次参数即可，其余地方调用不需要传参
function openScroll(option?: ScrollBarType): void {
	const payload = () =>
		option ? onSetProgressProp(option) : onSetProgressProp();
	onActivated(() => payload());
	onMounted(() => payload());
}

const closeScroll = (): void => {
	onDeactivated(() => onClear());
	onUnmounted(() => onClear());
};

const onClear = (): void => {
	if (timer) clearTimeout(timer);
	onUnSetProgressProp();
};

// 注册插件接受参数并挂载到全局
const useScroll = {
	install: (app: App, option?: ScrollBarType) => {
		const payload = () => (option ? openScroll(option) : openScroll());
		app.provide(scrollKey, {
			$openScroll: () => payload(),
			$closeScroll: () => closeScroll, // 不能在非setup上下文调用
		});

		app.config.globalProperties.$openScroll = () => payload();
		app.config.globalProperties.$closeScroll = closeScroll;
	},
};

export { useScroll, scrollKey };


lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.scrollKey = exports.useScroll = void 0;
var vue_1 = require("vue");
var timer = null;
// 声明provide的key用于inject接受类型注解
var scrollKey = Symbol();
exports.scrollKey = scrollKey;
function onUnSetProgressProp() {
    var _a, _b, _c;
    var el = document.querySelector(".ProgressTopBar");
    (_a = el === null || el === void 0 ? void 0 : el.style) === null || _a === void 0 ? void 0 : _a.removeProperty("background-image");
    (_b = el === null || el === void 0 ? void 0 : el.style) === null || _b === void 0 ? void 0 : _b.removeProperty("background-size");
    (_c = el === null || el === void 0 ? void 0 : el.style) === null || _c === void 0 ? void 0 : _c.removeProperty("background-repeat");
}
function onSetProgressProp(option) {
    if (option === void 0) { option = {
        progress: "red",
        progressRollback: "#fff",
        progressTop: "3px",
        progressLeft: "0px",
        UIViewBackground: "#fff",
    }; }
    timer = setTimeout(function () {
        var el = document.querySelector(".ProgressTopBar");
        if (el && el instanceof HTMLElement) {
            el.style.setProperty("position", "relative");
            el.style.setProperty("z-index", "1");
            el.style.setProperty("background-image", "linear-gradient(to right top,  ".concat(option.progress, " 50%, ").concat(option.progressRollback, " 50%)"));
            el.style.setProperty("background-size", "100% calc(100% - 100vh + ".concat(option.progressTop, " )"));
            el.style.setProperty("background-repeat", "no-repeat");
            //  无法设置CSS StyleDeclaration 的背景颜色属性：这些样式是计算出来的，因此背景颜色属性是只读的
            // console.log((getComputedStyle(el, '::after').backgroundColor = 'yellow'));
            var styleEl = document.createElement("style");
            styleEl.innerHTML = "\n\t\t\t\t.ProgressTopBar::after {\n\t\t\t\t\tcontent: '';\n\t\t\t\t\tposition: fixed;\n\t\t\t\t\ttop: ".concat(option.progressTop, ";\n\t\t\t\t\tleft: ").concat(option.progressLeft, ";\n\t\t\t\t\tbottom: 0;\n\t\t\t\t\tright: 0;\n\t\t\t\t\tbackground: ").concat(option.UIViewBackground, ";\n\t\t\t\t\tz-index: -1;\n\t\t\t\t}\n\t\t\t");
            document.head.appendChild(styleEl);
        }
    }, 500);
}
// 命名函数openScroll  参数可选、因为只需要在main.ts中传入一次参数即可，其余地方调用不需要传参
function openScroll(option) {
    var payload = function () {
        return option ? onSetProgressProp(option) : onSetProgressProp();
    };
    (0, vue_1.onActivated)(function () { return payload(); });
    (0, vue_1.onMounted)(function () { return payload(); });
}
var closeScroll = function () {
    (0, vue_1.onDeactivated)(function () { return onClear(); });
    (0, vue_1.onUnmounted)(function () { return onClear(); });
};
var onClear = function () {
    if (timer)
        clearTimeout(timer);
    onUnSetProgressProp();
};
// 注册插件接受参数并挂载到全局
var useScroll = {
    install: function (app, option) {
        var payload = function () { return (option ? openScroll(option) : openScroll()); };
        app.provide(scrollKey, {
            $openScroll: function () { return payload(); },
            $closeScroll: function () { return closeScroll; }, // 不能在非setup上下文调用
        });
        app.config.globalProperties.$openScroll = function () { return payload(); };
        app.config.globalProperties.$closeScroll = closeScroll;
    },
};
exports.useScroll = useScroll;
